# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/html/HTMLFormElement.cpp
# Commit: b9e87fde31f7
# Full Hash: b9e87fde31f7ca31185081dc96fa937927f1879f
# Author: sefeng <sefeng@mozilla.com>
# Date: 2020-05-12 04:04:10
# Regressor Bug: 1557303
# File Overlap Count: 1
# Description:
#   Bug 1557303 - Implement form[method="dialog"] r=smaug
#   
#   Spec: https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#submit-dialog
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D73364
# ==============================================================================

diff -r e02018dfce9e -r b9e87fde31f7 dom/html/HTMLFormElement.cpp
--- a/dom/html/HTMLFormElement.cpp	Tue May 12 00:11:12 2020 +0000
+++ b/dom/html/HTMLFormElement.cpp	Mon May 11 22:56:13 2020 +0000
@@ -338,6 +338,10 @@
                                      nsAttrValue& aResult) {
   if (aNamespaceID == kNameSpaceID_None) {
     if (aAttribute == nsGkAtoms::method) {
+      if (StaticPrefs::dom_dialog_element_enabled()) {
+        return aResult.ParseEnumValue(aValue, kFormMethodTableDialogEnabled,
+                                      false);
+      }
       return aResult.ParseEnumValue(aValue, kFormMethodTable, false);
     }
     if (aAttribute == nsGkAtoms::enctype) {
@@ -682,6 +686,26 @@
   //
   // perform the submission
   //
+  if (!submission) {
+    mIsSubmitting = false;
+#ifdef DEBUG
+    HTMLDialogElement* dialog = nullptr;
+    for (nsIContent* parent = GetParent(); parent;
+         parent = parent->GetParent()) {
+      dialog = HTMLDialogElement::FromNodeOrNull(parent);
+      if (dialog) {
+        break;
+      }
+    }
+    MOZ_ASSERT(!dialog || !dialog->Open());
+#endif
+    return NS_OK;
+  }
+
+  if (DialogFormSubmission* dialogSubmission =
+          submission->GetAsDialogSubmission()) {
+    return SubmitDialog(dialogSubmission);
+  }
   return SubmitSubmission(submission.get());
 }
 
@@ -726,8 +750,10 @@
   //
   // Dump the data into the submission object
   //
-  rv = formData->CopySubmissionDataTo(*aFormSubmission);
-  NS_ENSURE_SUBMIT_SUCCESS(rv);
+  if (!(*aFormSubmission)->GetAsDialogSubmission()) {
+    rv = formData->CopySubmissionDataTo(*aFormSubmission);
+    NS_ENSURE_SUBMIT_SUCCESS(rv);
+  }
 
   return NS_OK;
 }
@@ -841,6 +867,22 @@
   return rv;
 }
 
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#submit-dialog
+nsresult HTMLFormElement::SubmitDialog(DialogFormSubmission* aFormSubmission) {
+  mIsSubmitting = false;
+
+  // Close the dialog subject. If there is a result, let that be the return
+  // value.
+  HTMLDialogElement* dialog = aFormSubmission->DialogElement();
+  MOZ_ASSERT(dialog);
+
+  Optional<nsAString> retValue;
+  retValue = &aFormSubmission->ReturnValue();
+  dialog->Close(retValue);
+
+  return NS_OK;
+}
+
 nsresult HTMLFormElement::DoSecureToInsecureSubmitCheck(nsIURI* aActionURL,
                                                         bool* aCancelSubmit) {
   *aCancelSubmit = false;